src/middleware/axum.rs

```diff
@@ -0,0 +1,209 @@
+    S: Service<Request<B>, Response = Response<B2>> + Clone + Send + 'static,
+    S::Error: Error + 'static, //fmt::Display + 'static,
+    S::Future: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    // #[allow(clippy::type_complexity)]
+    // type Future = futures_core::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<B>) -> Self::Future {
+        use tracing_opentelemetry::OpenTelemetrySpanExt;
+        let req = req;
+        let span = if self.filter.map_or(true, |f| f(req.uri().path())) {
+            let span = otel_http::http_server::make_span_from_request(&req);
+            let route = http_route(&req);
+            let method = otel_http::http_method(req.method());
+            // let client_ip = parse_x_forwarded_for(req.headers())
+            //     .or_else(|| {
+            //         req.extensions()
+            //             .get::<ConnectInfo<SocketAddr>>()
+            //             .map(|ConnectInfo(client_ip)| Cow::from(client_ip.to_string()))
+            //     })
+            //     .unwrap_or_default();
+            span.record("http.route", route);
+            span.record("otel.name", format!("{method} {route}").trim());
+            // span.record("trace_id", find_trace_id_from_tracing(&span));
+            // span.record("client.address", client_ip);
+            span.set_parent(otel_http::extract_context(req.headers()));
+            span
+        } else {
+            tracing::Span::none()
+        };
+        let future = {
+            let _ = span.enter();
+            self.inner.call(req)
+        };
+        ResponseFuture {
+            inner: future,
+            span,
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`Trace`].
+    ///
+    /// [`Trace`]: super::Trace
+    pub struct ResponseFuture<F> {
+        #[pin]
+        pub(crate) inner: F,
+        pub(crate) span: Span,
+        // pub(crate) start: Instant,
+    }
+}
+
+impl<Fut, ResBody, E> Future for ResponseFuture<Fut>
+where
+    Fut: Future<Output = Result<Response<ResBody>, E>>,
+    E: std::error::Error + 'static,
+{
+    type Output = Result<Response<ResBody>, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let _guard = this.span.enter();
+        let result = futures_util::ready!(this.inner.poll(cx));
+        otel_http::http_server::update_span_from_response_or_error(this.span, &result);
+        Poll::Ready(result)
+    }
+}
+
+#[inline]
+fn http_route<B>(req: &Request<B>) -> &str {
+    req.extensions()
+        .get::<MatchedPath>()
+        .map_or_else(|| "", |mp| mp.as_str())
+}
+
+// #[cfg(test)]
+// mod tests {
```

Comment:
Are you planning to uncomment and fix those tests later?

If not, then I would rather just delete them.